File: Python/Containers/MongoDB-REST/mongodb_rest_client.py
import json
from bson import ObjectId
from typing import Optional, Any

from mongodb_crud_client import MongoDB_CRUD_Client

class MongoDB_REST_Client:
    def __init__(self, mongodb_crud_client: MongoDB_CRUD_Client):
        self.__crud_client = mongodb_crud_client

    def __bad_request(self, message: str) -> dict:
        return {
            'statusCode': 400,
            'headers': {'Content-Type': 'text/plain'},
            'body': json.dumps(message)
        }

    def __error_message(self, e: Exception) -> dict:
        return {
            'statusCode': 500,
            'headers': {'Content-Type': 'text/plain'},
            'body': str(e)
        }

    def __not_found(self, id: ObjectId) -> dict:
        return {
            'statusCode': 404,
            'headers': {'Content-Type': 'text/plain'},
            'body': f'Error: Document with id {id} not found'
        }

    def __ok(self, data: dict) -> dict:
        return {
            'statusCode': 200,
            'headers': {'Content-Type': 'application/json'},
            'body': json.dumps(data, default=str)
        }
    
    def __verify_parameters(self, db_name: Optional[str], collection_name: Optional[str], id: Optional[str], is_id_required: bool, data: Optional[Any], is_data_required: bool) -> Optional[dict]:
        if db_name is None:
            return self.__bad_request('Error: Missing db_name parameter')
        if collection_name is None:
            return self.__bad_request('Error: Missing collection_name parameter')
        if is_id_required:
            if id is None:
                return self.__bad_request('Error: Missing id parameter')
            elif not ObjectId.is_valid(id):
                return self.__bad_request('Error: Invalid id')
        if is_data_required and data is None:
            return self.__bad_request('Error: Missing body')
        return None
    
    def ping(self):
        self.__crud_client.ping()

    def get(self, db_name: Optional[str], collection_name: Optional[str], id: Optional[str]) -> dict: 
        try:
            error = self.__verify_parameters(db_name, collection_name, id, True, None, False)
            if error is not None:
                return error
            
            document = self.__crud_client.read(db_name, collection_name, ObjectId(id))
            if document is not None:
                return self.__ok(document)
            return self.__not_found(id)
        except Exception as e:
            return self.__error_message(e)
        
    def post(self, db_name: Optional[str], collection_name: Optional[str], data: Optional[any]) -> dict:
        try:
            error = self.__verify_parameters(db_name, collection_name, None, False, data, True)
            if error is not None:
                return error
            
            document_id = self.__crud_client.create(db_name, collection_name, data)
            if document_id is not None:
                return {
                    'statusCode': 201,
                    'body': str(document_id)
                }
            return {
                'statusCode': 500,
                'body': 'Failed to create document'
            }
        except Exception as e:
            return self.__error_message(e)

    def put(self, db_name: Optional[str], collection_name: Optional[str], id: Optional[str], data: Optional[any]) -> dict: 
        try:
            error = self.__verify_parameters(db_name, collection_name, id, True, data, True)
            if error is not None:
                return error
            
            document = self.__crud_client.update(db_name, collection_name, ObjectId(id), data)
            if document is not None:
                return self.__ok(document)
            return self.__not_found(id)
        except Exception as e:
            return self.__error_message(e)

    def delete(self, db_name: Optional[str], collection_name: Optional[str], id: Optional[str]) -> dict: 
        try:
            error = self.__verify_parameters(db_name, collection_name, id, True, None, False)
            if error is not None:
                return error
            
            result = self.__crud_client.delete(db_name, collection_name, ObjectId(id))
            if result:
                return {
                    'statusCode': 200,
                    'body': f'Document with id {id} deleted'
                }
            return self.__not_found(id)
        except Exception as e:
            return self.__error_message(e)

File: Examples/PHP/MongoDB/REST/MongoDB_REST_Client.php
<?php

require_once '../CRUD/MongoDB_CRUD_Client.php';

class MongoDB_REST_Client {
    private $crud_client;

    public function __construct($mongodb_crud_client) {
        $this->crud_client = $mongodb_crud_client;
    }

    private function bad_request($message) {
        return [
            'statusCode' => 400,
            'headers' => ['Content-Type' => 'text/plain'],
            'body' => json_encode($message)
        ];
    }

    private function error_message($e) {
        return [
            'statusCode' => 500,
            'headers' => ['Content-Type' => 'text/plain'],
            'body' => $e->getMessage()
        ];
    }

    private function not_found($id) {
        return [
            'statusCode' => 404,
            'headers' => ['Content-Type' => 'text/plain'],
            'body' => "Error: Document with id $id not found"
        ];
    }

    private function ok($data) {
        return [
            'statusCode' => 200,
            'headers' => ['Content-Type' => 'application/json'],
            'body' => json_encode($data)
        ];
    }

    private function verify_parameters($db_name, $collection_name, $id, $is_id_required, $data, $is_data_required) {
        if ($db_name === null) {
            return $this->bad_request('Error: Missing db_name parameter');
        }
        if ($collection_name === null) {
            return $this->bad_request('Error: Missing collection_name parameter');
        }
        if ($is_id_required) {
            if ($id === null) {
                return $this->bad_request('Error: Missing id parameter');
            } elseif (!MongoDB\BSON\ObjectId::isValid($id)) {
                return $this->bad_request('Error: Invalid id');
            }
        }
        if ($is_data_required && $data === null) {
            return $this->bad_request('Error: Missing body');
        }
        return null;
    }

    public function ping() {
        $this->crud_client->ping();
    }

    public function get($db_name, $collection_name, $id) {
        try {
            $error = $this->verify_parameters($db_name, $collection_name, $id, true, null, false);
            if ($error !== null) {
                return $error;
            }

            $document = $this->crud_client->read($db_name, $collection_name, new MongoDB\BSON\ObjectId($id));
            if ($document !== null) {
                return $this->ok($document);
            }
            return $this->not_found($id);
        } catch (Exception $e) {
            return $this->error_message($e);
        }
    }

    public function post($db_name, $collection_name, $data) {
        try {
            $error = $this->verify_parameters($db_name, $collection_name, null, false, $data, true);
            if ($error !== null) {
                return $error;
            }

            $document_id = $this->crud_client->create($db_name, $collection_name, $data);
            if ($document_id !== null) {
                return [
                    'statusCode' => 201,
                    'body' => (string)$document_id
                ];
            }
            return [
                'statusCode' => 500,
                'body' => 'Failed to create document'
            ];
        } catch (Exception $e) {
            return $this->error_message($e);
        }
    }

    public function put($db_name, $collection_name, $id, $data) {
        try {
            $error = $this->verify_parameters($db_name, $collection_name, $id, true, $data, true);
            if ($error !== null) {
                return $error;
            }

            $document = $this->crud_client->update($db_name, $collection_name, new MongoDB\BSON\ObjectId($id), $data);
            if ($document !== null) {
                return $this->ok($document);
            }
            return $this->not_found($id);
        } catch (Exception $e) {
            return $this->error_message($e);
        }
    }

    public function delete($db_name, $collection_name, $id) {
        try {
            $error = $this->verify_parameters($db_name, $collection_name, $id, true, null, false);
            if ($error !== null) {
                return $error;
            }

            $result = $this->crud_client->delete($db_name, $collection_name, new MongoDB\BSON\ObjectId($id));
            if ($result) {
                return [
                    'statusCode' => 200,
                    'body' => "Document with id $id deleted"
                ];
            }
            return $this->not_found($id);
        } catch (Exception $e) {
            return $this->error_message($e);
        }
    }
}